### PR TITLE
Impl transformers adam w

### PR DESCRIPTION
**Pull request by et16kr**
- Added hugging face transformers adam optimizer.
```from transformers import AdamW```
- This is the fused version of [AdamW included in the Hugging Face Transformers module](https://github.com/huggingface/transformers/blob/1c52cb7b3b194bf1ad2cf6ceaee698a12689b9fa/src/transformers/optimization.py#L607).
- Implements Adam algorithm with weight decay fix as introduced in [Decoupled Weight Decay Regularization](https://arxiv.org/abs/1711.05101).
- Added driver test and gtest for transformers_adam_w.
- I couldn't find a suitable short name. If you suggest a name, I will change it.

<details><summary>adam float16</summary>
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

dims | rocm pytorch | miopen | imporvement | model
-- | -- | -- | -- | --
96 | 25339 | 3984 | 6.36 | gpt,gpt2
288 | 29882 | 4368 | 6.84 | gpt2
768 | 33130 | 4736 | 7.00 | gpt,gpt2
2304 | 38728 | 5232 | 7.40 | gpt,gpt2
3072 | 38696 | 5488 | 7.05 | gpt,gpt2
96x3072 | 49446 | 9248 | 5.35 | gpt
512x768 | 50022 | 9952 | 5.03 | gpt
768x768 | 53653 | 13137 | 4.08 | gpt,gpt2
768x2304 | 96717 | 30433 | 3.18 | gpt,gpt2
768x3072 | 112057 | 38114 | 2.94 | gpt,gpt2
1024x768 | 56548 | 16144 | 3.50 | gpt2
3072x768 | 112027 | 37810 | 2.96 | gpt,gpt2
52000x768 | 1422900 | 703501 | 2.02 | gpt,gpt2

</body>

</html>

</details>

<details><summary> adam float32</summary>
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

dims | rocm pytorch | miopen | imporvement | model
-- | -- | -- | -- | --
96 | 23868 | 4288 | 5.57 | gpt,gpt2
288 | 26858 | 4608 | 5.83 | gpt2
768 | 29322 | 5456 | 5.37 | gpt,gpt2
2304 | 34937 | 6016 | 5.81 | gpt,gpt2
3072 | 35417 | 6032 | 5.87 | gpt,gpt2
96x3072 | 48726 | 11392 | 4.28 | gpt
512x768 | 51510 | 13392 | 3.85 | gpt
768x768 | 59636 | 16641 | 3.58 | gpt,gpt2
768x2304 | 124871 | 38130 | 3.27 | gpt,gpt2
768x3072 | 152386 | 49586 | 3.07 | gpt,gpt2
1024x768 | 70066 | 20241 | 3.46 | gpt2
3072x768 | 152610 | 49570 | 3.08 | gpt,gpt2
52000x768 | 2173598 | 856019 | 2.54 | gpt,gpt2

</body>

</html>

</details>

<details><summary>amp adam (parameter float32, gradient float16)</summary>
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

dims | rocm pytorch | miopen | imporvement | model
-- | -- | -- | -- | --
96 | 25915 | 9936 | 2.61 | gpt,gpt2
288 | 24619 | 10065 | 2.45 | gpt2
768 | 29002 | 10960 | 2.65 | gpt,gpt2
2304 | 34953 | 11696 | 2.99 | gpt,gpt2
3072 | 34937 | 11537 | 3.03 | gpt,gpt2
96x3072 | 48630 | 17217 | 2.82 | gpt
512x768 | 51398 | 18705 | 2.75 | gpt
768x768 | 59620 | 22417 | 2.66 | gpt,gpt2
768x2304 | 124168 | 43794 | 2.84 | gpt,gpt2
768x3072 | 152466 | 55666 | 2.74 | gpt,gpt2
1024x768 | 69986 | 27345 | 2.56 | gpt2
3072x768 | 152498 | 55410 | 2.75 | gpt,gpt2
52000x768 | 2168876 | 907477 | 2.39 | gpt,gpt2

</body>

</html>

</details>

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/et16/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">

</head>

<body link="#0563C1" vlink="#954F72">

type | avrage
-- | --
adam float16 | 4.28
adam float32 | 4.90
amp adam | 2.71

</body>

</html>

</details>

### CODE REVIEW

**Comment by sangwon-chae** [comment]
@JehandadKhan @junliume Would you please review this PR?

**Comment by junliume** [comment]
Add @CAHEK7 @DrizztDoUrden @amberhassaan @atamazov for second review

**Review by junliume — APPROVED** [review]

File: include/miopen/miopen.h [hunk 1]
```diff
@@ -7211,6 +7211,286 @@ miopenFusedAdamWithOutput(miopenHandle_t handle,
...
+ * @param correct_bias        Whether or not to correct bias in Adam (for instance, in Bert TF
+ *                            repository they use False).
+ * @param gradScaleDesc       Tensor descriptor for the input grad scale tensor (input, optional)
+ * @param gradScale           Input grad scale tensor (input, optional)
+ * @param foundInfDesc        Tensor descriptor for the input found inf tensor (input, optional)
+ * @param foundInf            Tensor indicating the presence of inf or NaN in gradients. If true,
+ *                            skips operation and step update (input, optional)
+ * @return                    miopenStatus_t
+ */
+MIOPEN_EXPORT miopenStatus_t miopenTransformersAdamW(miopenHandle_t handle,
```
Comment (atamazov) [inline include/miopen/miopen.h:7352]: We already have `adamw` parameter in `miopenFusedAdam()`, marked as "not supported". Do we really need to add new API calls? Shall we better extend the existing fused Adam stuff?

File: driver/transformers_adam_w_driver.hpp
Comment (atamazov) [inline driver/transformers_adam_w_driver.hpp:1]: The existing adam_driver.hpp is very similar to this. The only differences are;
- mloAdamRunHost (but the code in the new file is very similar again)
- is_amp parameter
- handling of maxExpAvgSqDesc
- correct_bias

Shall we generalize existing driver module instead of introducing a new one?

File: src/tensor.cpp [hunk 2]
```diff
@@ -458,6 +458,22 @@ std::size_t TensorDescriptor::GetNumBytes() const
...
+    size_t plane_size    = 1;
+    size_t dims_of_shape = lens.size();
+
+    for(int index = dims_of_shape - 1; index >= 0; --index)
+    {
+        if((lens[index] != 1) && (strides[index] != plane_size))
+        {
+            return false;
+        }
+        plane_size *= static_cast<int64_t>(lens[index]);
```
Comment (atamazov) [inline src/tensor.cpp]: Both plane_size and lens[x] are of the size_t type. Why static_cast?

File: src/tensor.cpp [hunk 3]
```diff
@@ -458,6 +458,22 @@ std::size_t TensorDescriptor::GetNumBytes() const
 
 bool TensorDescriptor::IsPacked() const { return this->packed; }
 
+bool TensorDescriptor::IsContiguous() const
```
Comment (atamazov) [inline src/tensor.cpp]: Please add a comment that explains the physical meaning of this, i.e. what is an contiguous tensor. What is the difference from "packed"? And, if this check is Adam-specific, then please move this check to the adam solver.

File: src/adam.cpp [hunk 4]
```diff
@@ -89,19 +89,13 @@ miopenStatus_t Adam(Handle& handle,
...
-                                                  eps,
                                                   amsgrad,
-                                                  maximize,
+                                                  true,
                                                   adamw,
                                                   is_amp};
 
     const auto invoke_params = [&]() {
-        auto tmp = adam::InvokeParams{};
+        auto tmp = adam::AdamInvokeParams{};
```
Comment (atamazov) [inline src/adam.cpp:98]: [Q] Why this rename is necessary? `adam::InvokeParams` looks good to me.

UPDATE: Now I see. Let's use the same InvokeParams for Adam and AdamW. You can use union to minimize footprint of InvokeParams, if you wish (up to you).

File: src/include/miopen/adam.hpp [hunk 5]
```diff
@@ -73,5 +73,39 @@ MIOPEN_INTERNALS_EXPORT miopenStatus_t Adam(Handle& handle,
                                             bool adamw,
                                             bool is_amp);
 
+miopenStatus_t TransformersAdamW(Handle& handle,
```
Comment (atamazov) [inline src/include/miopen/adam.hpp]: MIOPEN_INTERNALS_EXPORT seems missing.

File: src/include/miopen/adam/problem_description.hpp [hunk 6]
```diff
@@ -77,14 +71,8 @@ struct MIOPEN_INTERNALS_EXPORT ProblemDescription : ProblemDescriptionBase
...
           stepOutDesc(stepOutDesc_),
-          step(step_),
-          lr(lr_),
-          beta1(beta1_),
-          beta2(beta2_),
-          weight_decay(weight_decay_),
-          eps(eps_),
           amsgrad(amsgrad_),
-          maximize(maximize_),
+          correct_bias(correct_bias_),
```
Comment (atamazov) [inline src/include/miopen/adam/problem_description.hpp:75]: Yeah, now I see that the same Problem is used for both Adam and AdamW. Which means that these two are basically the same primitive. So let's add one more Solver to the existing Adam solvers, make sure that IsApplicable in the old Adam solver is updated (these should return false for the "adamw" Adam problems) and let the core of MIOpen to do the rest (i.e. select the proper Solver).

**Review by atamazov — CHANGES_REQUESTED** [review]
It's sad to say, but I'm getting more(h) and more(h) upset that patches contain dupes of the existing code (but maybe I am wrong?) This is a quality defect that would not be easy to detect and fix in the future .

File: include/miopen/miopen.h (hunk 1 above)
Comment (et16kr) [reply include/miopen/miopen.h:7352]: This PR is for the AdamW included in Huggingface Transformers.
I submitted PR https://github.com/ROCm/MIOpen/pull/2957 for AdamW in PyTorch using the adamw flag in miopenFusedAdam().
miopenTransformersAdamW and miopenFusedAdam have different internal calculations and API function parameters. I submitted each PR separately.

File: src/tensor.cpp (hunk 3 above)
Comment (et16kr) [reply src/tensor.cpp]: IsPacked() is only False when GetElementSize() != GetElementSpace(). 
Next example, tensor is packed and not contiguous.
```
>>> tensor = torch.tensor([[1,2,3,4],[5,6,7,8]])
>>> tensor = torch.as_strided(tensor, size=tensor.size(), stride=(1,2))
>>> tensor
tensor([[1, 3, 5, 7],
        [2, 4, 6, 8]])
>>> tensor.is_contiguous()
False
```
And this diff is taken from https://github.com/ROCm/MIOpen/pull/2970.
If https://github.com/ROCm/MIOpen/pull/2970 is merged first, it will be deleted from this PR.

File: src/tensor.cpp (hunk 2 above)
Comment (et16kr) [reply src/tensor.cpp]: I fixed it.

File: src/include/miopen/adam.hpp (hunk 5 above)
Comment (et16kr) [reply src/include/miopen/adam.hpp]: I fixed it.

File: src/include/miopen/adam/problem_description.hpp (hunk 6 above)
Comment (et16kr) [reply src/include/miopen/adam/problem_description.hpp:75]: If combining into one solver, it would be better if https://github.com/ROCm/MIOpen/pull/2957 is reviewed and merged first.

**Comment by et16kr** [comment]
[#2957 AdamW](https://github.com/ROCm/MIOpen/pull/2957) has been merged.
I have resolved the merge conflict.
I will now begin the integration with Adam.
